elfo-core/src/message/repr.rs

```diff
@@ -0,0 +1,340 @@
+use std::{
+    alloc, fmt,
+    ptr::{self, NonNull},
+};
+
+use fxhash::{FxHashMap, FxHashSet};
+use metrics::Label;
+use once_cell::sync::Lazy;
+use smallbox::smallbox;
+
+use super::Message;
+use crate::dumping;
+
+#[cfg(feature = "network")]
+use rmp_serde::{decode, encode};
+
+// === MessageTypeId ===
+
+/// A unique (inside a compilation target) identifier of a message type.
+// Internally, it's simply an address of corresponding vtable.
+// ~
+// However, we cannot cast it into integer in the const context,
+// so we're forced to use a raw pointer and `ptr::eq()`.
+//
+// `NULL` is used for `AnyMessage`.
+//
+// Reexported in `elfo::_priv`.
+#[doc(hidden)]
+#[derive(Clone, Copy, Debug)]
+pub struct MessageTypeId(*const ());
+
+/// SAFETY: used only for comparison, safe to send across threads.
+unsafe impl Send for MessageTypeId {}
+/// SAFETY: used only for comparison, safe to sync across threads.
+unsafe impl Sync for MessageTypeId {}
+
+impl MessageTypeId {
+    #[inline]
+    pub const fn new(vtable: &'static MessageVTable) -> Self {
+        Self(vtable as *const _ as *const ())
+    }
+
+    pub(super) const fn any() -> Self {
+        Self(ptr::null())
+    }
+}
+
+impl PartialEq for MessageTypeId {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        ptr::eq(self.0, other.0)
+    }
+}
+
+// === MessageRepr ===
+
+/// A message representation as a cpp-style object.
+///
+/// Initially, it's created from a typed message as [`MessageRepr<M>`], then
+/// * for [`AnyMessage`]: moved directly to heap
+/// * for [`Envelope`]: becomes a part and whole envelope is moved to heap
+///
+/// All subsequent accesses are done via `NonNull<MessageRepr>` and require
+/// using the virtual table ([`MessageVTable`]) for all operations.
+///
+/// [`AnyMessage`]: crate::AnyMessage
+/// [`Envelope`]: crate::Envelope
+// `vtable` must be first
+#[repr(C)]
+// It's `pub` only because used in private methods of `AnyMessage`.
+// Actually, it's not reexported at all.
+#[doc(hidden)]
+pub struct MessageRepr<M = ()> {
```

Comment:
If I understood correctly, `M = ()` stands for “any message”, kind of like `void*`. Maybe a type with more descriptive name could be used here instead of `()`?